sklearn/semi_supervised/tests/test_self_training.py

```diff
@@ -0,0 +1,165 @@
+
+            assert_array_equal(np.vectorize(mapping.get)(pred), pred_string)
+            assert_array_equal(proba, proba_string)
+
+
+def test_missing_predict_proba():
+    # Check that an error is thrown if predict_proba is not implemented
+    base_classifier = SVC(gamma="scale")
+    self_training = SelfTrainingClassifier(base_classifier)
+    message = "base_classifier (SVC) should implement predict_proba!"
+    assert_raise_message(ValueError, message, self_training.fit, X_train,
+                         y_train)
+
+
+def test_invalid_params():
+    # Test negative iterations
+    grid = ParameterGrid({"max_iter": [-1, -100, -10]})
+    base_classifier = SVC(gamma="scale", probability=True)
+    for params in grid:
+        st = SelfTrainingClassifier(base_classifier, **params)
+        message = "max_iter must be >= 0"
+        assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+    grid = ParameterGrid({"threshold": [1.0, -2, 10]})
+    base_classifier = SVC(gamma="scale", probability=True)
+    for params in grid:
+        st = SelfTrainingClassifier(base_classifier, **params)
+        message = "threshold must be in [0,1)"
+        assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+
+def test_single_iteration():
+    # Check classification for single iteration.
+    # Fitting a SelfTrainingClassifier with one iteration and 100 unlabeled
+    # datapoints should give the same results as fitting a normal classifier
+    # with only 50 labeled datapoints.
+
+    clf1 = SelfTrainingClassifier(KNeighborsClassifier(),
+                                  max_iter=0).fit(X_train,
+                                                  y_train_missing_labels)
+
+    clf2 = KNeighborsClassifier().fit(X_train[:limit], y_train[:limit])
+
+    assert_array_equal(clf1.predict(X_test), clf2.predict(X_test))
+
+
+def test_notfitted():
+    # Test that predicting without training throws an error
+    st = SelfTrainingClassifier(KNeighborsClassifier())
+    msg = ("This SelfTrainingClassifier instance is not fitted yet. Call "
+           "\'fit\' with appropriate arguments before using this method.")
+    assert_raise_message(NotFittedError, msg, st.predict, X_train)
+    assert_raise_message(NotFittedError, msg, st.predict_proba, X_train)
+
+
+def test_y_labeled_iter():
+    # Check that the amount of datapoints labeled in iteration 0 is equal to
+    # the amount of labeled datapoints we passed.
+    for m in range(1, 5):
+        st = SelfTrainingClassifier(KNeighborsClassifier(), max_iter=m)
+        st.fit(X_train, y_train_missing_labels)
+        amount_iter_0 = len(st.y_labeled_iter_[st.y_labeled_iter_ == 0])
+        assert(amount_iter_0 == 50)
+        # Check that the max of the iterations is less than the total amount of
+        # iterations
+        assert(np.max(st.y_labeled_iter_) <= m)
+        assert(np.max(st.y_labeled_iter_) <= st.n_iter_)
+
+
+def test_prefitted_throws_error():
```

Comment:
Put this next to `test_notfitted` and maybe merge them as they're testing the same thing